format/postgres/flavours/postgres14/common14/pgheap.go

```diff
@@ -0,0 +1,492 @@
+	d.SeekAbs(0)
+	decodeHeapPages(d)
+
+	return nil
+}
+
+func decodeHeapPages(d *decode.D) {
+	heap := GetHeapD(d)
+
+	for {
+		if end, _ := d.TryEnd(); end {
+			return
+		}
+
+		page := &HeapPageD{}
+		heap.Page = page
+
+		d.FieldStruct("HeapPage", decodeHeapPage)
+
+		// end of Page
+		endLen := uint64(d.Pos() / 8)
+		pageEnd := common.TypeAlign(heap.PageSize, endLen)
+		d.SeekAbs(int64(pageEnd) * 8)
+	}
+}
+
+func decodeHeapPage(d *decode.D) {
+	heap := GetHeapD(d)
+
+	page := &HeapPageD{}
+	heap.Page = page
+
+	pagePosBegin := common.RoundDown(heap.PageSize, uint64(d.Pos()/8))
+	page.PagePosBegin = pagePosBegin
+
+	// PageHeader
+	d.FieldStruct("PageHeaderData", heap.DecodePageHeaderDataFn)
+
+	// free space
+	freeSpaceEnd := int64(pagePosBegin*8) + int64(page.PdUpper*8)
+	freeSpaceNBits := freeSpaceEnd - d.Pos()
+	if freeSpaceNBits != 0 {
+		d.FieldRawLen("FreeSpace", freeSpaceNBits, scalar.RawHex)
+	}
+
+	if uint64(page.PdSpecial) != heap.PageSize && heap.DecodePageSpecialFn != nil {
+		heap.DecodePageSpecialFn(d)
+	}
+
+	// Tuples
+	d.FieldArray("Tuples", decodeTuples)
+}
+
+/*    0      |     8 */ // PageXLogRecPtr pd_lsn;
+/*    8      |     2 */ // uint16 pd_checksum;
+/*   10      |     2 */ // uint16 pd_flags;
+/*   12      |     2 */ // LocationIndex pd_lower;
+/*   14      |     2 */ // LocationIndex pd_upper;
+/*   16      |     2 */ // LocationIndex pd_special;
+/*   18      |     2 */ // uint16 PdPagesizeVersion;
+/*   20      |     4 */ // TransactionId pd_prune_xid;
+/*   24      |     0 */ // ItemIdData pd_linp[];
+func DecodePageHeaderData(d *decode.D) {
+	heap := GetHeapD(d)
+	page := heap.Page
+
+	d.FieldStruct("pd_lsn", func(d *decode.D) {
+		/*    0      |     4 */ // uint32 xlogid;
+		/*    4      |     4 */ // uint32 xrecoff;
+		d.FieldU32("xlogid", common.HexMapper)
+		d.FieldU32("xrecoff", common.HexMapper)
+	})
+	d.FieldU16("pd_checksum")
+	d.FieldU16("pd_flags")
+	page.PdLower = uint16(d.FieldU16("pd_lower"))
+	page.PdUpper = uint16(d.FieldU16("pd_upper"))
+	page.PdSpecial = uint16(d.FieldU16("pd_special"))
+	page.PdPagesizeVersion = uint16(d.FieldU16("pd_pagesize_version"))
+	d.FieldU32("pd_prune_xid")
+
+	// ItemIdData pd_linp[];
+	page.ItemsEnd = int64(page.PagePosBegin*8) + int64(page.PdLower*8)
+	d.FieldArray("pd_linp", DecodeItemIds)
+}
+
+func DecodeItemIds(d *decode.D) {
+	heap := GetHeapD(d)
+	page := heap.Page
+
+	for {
+		checkPos := d.Pos()
+		if checkPos >= page.ItemsEnd {
+			break
+		}
+		/*    0: 0   |     4 */ // unsigned int lp_off: 15
+		/*    1: 7   |     4 */ // unsigned int lp_flags: 2
+		/*    2: 1   |     4 */ // unsigned int lp_len: 15
+		d.FieldStruct("ItemIdData", func(d *decode.D) {
+			itemId := itemIdDataD{}
+
+			itemPos := d.Pos()
+			itemId.lpOff = uint32(d.FieldU32("lp_off", common.LpOffMapper))
+			d.SeekAbs(itemPos)
+			itemId.lpFlags = uint32(d.FieldU32("lp_flags", common.LpFlagsMapper))
+			d.SeekAbs(itemPos)
+			itemId.lpLen = uint32(d.FieldU32("lp_len", common.LpLenMapper))
```

Comment:
Is there some reason this can't be?
```
itemId.lpOff = d.FieldU15("lp_off")
itemId.lpFlags = d.FieldU2("lp_off")
itemId.lpLen = d.FieldU15("lp_off")
```
Oh is it little endian bit-ranges the make things messy?

Reply:
Sorry, but this not working:
```
itemId.lpOff = d.FieldU15("lp_off")
itemId.lpFlags = d.FieldU2("lp_off")
itemId.lpLen = d.FieldU15("lp_off")
```
I was trying. And checked again.

Originally in C it was:
```
typedef struct ItemIdData
{
	unsigned	lp_off:15,		/* offset to tuple (from start of page) */
                        lp_flags:2,		/* state of line pointer, see below */
                        lp_len:15;		/* byte length of tuple */
} ItemIdData;
```
We need to use bit shifting to get correct values:
```
v := (s.ActualU() >> 17) & 0x7fff
```

Reply:
Aha i see, it's 32 bit little endian value where the bit ranges span bytes? that is awkward to decode and present with fq at the moment. I've ran into this before with vp8_frame https://github.com/wader/fq/blob/master/format/vpx/vp8_frame.go#L56
It would be nice to come up both with some generic decode helper to make it easier and also some way of presenting it but i have no ideas at the moment.